lib/Backends/OpenCL/OpenCLDeviceManager.cpp

```diff
@@ -181,6 +181,52 @@ Error OpenCLDeviceManager::parseConfig() {
   return Error::success();
 }
 
+Error OpenCLDeviceManager::findBestDevice(cl_platform_id platformId,
+                                          int deviceId) {
+  cl_uint num{0};
+  cl_int err = clGetDeviceIDs(platformId, CL_DEVICE_TYPE_ALL, 0, nullptr, &num);
+  if (err != CL_SUCCESS) {
+    RETURN_ERR("clGetDeviceIDs Failed");
+  }
+  if ((deviceId > 0 && num < deviceId) || num == 0) {
+    RETURN_ERR("Should have at least one GPU/CPU/FPGA for running OpenCL");
+  }
+
+  std::vector<cl_device_id> devices(num);
+  err = clGetDeviceIDs(platformId, CL_DEVICE_TYPE_ALL, num, devices.data(),
+                       nullptr);
+  if (err != CL_SUCCESS) {
+    RETURN_ERR("clGetDeviceIDs Failed");
+  }
+
+  if (deviceId >= 0) {
+    deviceId_ = devices[deviceId];
+  } else {
+    cl_device_id chosen = devices[0];
+    cl_device_type chosen_type = 0;
+
+    for (auto id : devices) {
+      cl_bitfield type;
+      err = clGetDeviceInfo(id, CL_DEVICE_TYPE, sizeof(cl_bitfield), &type,
+                            nullptr);
+      if (err != CL_SUCCESS) {
+        RETURN_ERR("clGetDeviceInfo Failed");
+      }
+
+      if (type >= chosen_type) {
+        chosen = id;
+      }
+    }
+    deviceId_ = chosen;
+  }
+
+  char name[100];
+  err = clGetDeviceInfo(deviceId_, CL_DEVICE_NAME, 100, &name, nullptr);
+  llvm::dbgs() << "Using OpenCL device " << name << "\n";
```

Comment:
Does it need to be printed unconditionally? Maybe it should be DEBUG_GLOW or something like that?